src/MultiFormats/VarInt.cs

```diff
@@ -0,0 +1,287 @@
+        /// <param name="stream">
+        ///   A varint encoded <see cref="Stream"/>.
+        /// </param>
+        /// <param name="cancel">
+        ///   Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException"/> is raised.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation. The task's result
+        ///   is the integer value in the <paramref name="stream"/>.
+        /// </returns>
+        /// <exception cref="EndOfStreamException">
+        ///   When the no bytes exist in the <paramref name="stream"/>.
+        /// </exception>
+        /// <exception cref="InvalidDataException">
+        ///   When the varint value is greater than <see cref="Int32.MaxValue"/>.
+        /// </exception>
+        public static async Task<int> ReadVarint32Async(this Stream stream,
+            CancellationToken cancel = default)
+        {
+            var value = await stream.ReadVarint64Async(cancel).ConfigureAwait(false);
+            if (value > int.MaxValue)
+                throw new InvalidDataException("Varint value is bigger than an Int32.MaxValue");
+            return (int) value;
+        }
+
+        /// <summary>
+        ///   Reads a variable integer from the stream. 
+        /// </summary>
+        /// <param name="stream">
+        ///   A varint encoded <see cref="Stream"/>.
+        /// </param>
+        /// <param name="cancel">
+        ///   Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException"/> is raised.
+        /// </param>
+        /// <exception cref="InvalidDataException">
+        ///   When the varint value is greater than <see cref="Int64.MaxValue"/>.
+        /// </exception>
+        /// <exception cref="EndOfStreamException">
+        ///   When the no bytes exist in the <paramref name="stream"/>.
+        /// </exception>
+        /// <returns>
+        ///   A task that represents the asynchronous operation. The task's result
+        ///   is the integer value in the <paramref name="stream"/>.
+        /// </returns>
+        public static async Task<long> ReadVarint64Async(this Stream stream,
+            CancellationToken cancel = default)
+        {
+            long value = 0;
+            var shift = 0;
+            var bytesRead = 0;
+            var buffer = new byte[1];
+
+            while (true)
+            {
+                if (1 != await stream.ReadAsync(buffer, 0, 1, cancel).ConfigureAwait(false))
+                {
+                    if (bytesRead == 0)
+                        throw new EndOfStreamException();
+
+                    throw new InvalidDataException("Varint is not terminated");
+                }
+
+                if (++bytesRead > 9) throw new InvalidDataException("Varint value is bigger than an Int64.MaxValue");
+
+                var b = buffer[0];
+                value |= (long) (b & 0x7F) << shift;
+
+                if (b < 0x80) return value;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)